### PR TITLE
[SPMD] Enable fused Adam in full train step tracing

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98113
* #98114

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

### CODE REVIEW

File: torch/distributed/_tensor/dispatch.py [hunk 1]
```diff
@@ -25,7 +25,7 @@
 
 
 def wrap(res: object, spec: OutputSpecType) -> object:
-    if isinstance(res, torch.Tensor):
+    def to_dt(res, spec):
```
Comment (mrshenli) [inline torch/distributed/_tensor/dispatch.py:28]: Hey @wanchaol , is this change OK? fused adam returns tuple of lists, so we need this.

**Review by yifuwang — APPROVED** [review]
This looks great! Thank you!

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/).

File: torch/distributed/_spmd/distribute.py [hunk 2]
```diff
@@ -388,8 +388,13 @@ def _rebuild_graph(
...
                             isinstance(n.target, torch._ops.OpOverload)
-                            and n.target._schema.name.startswith(
-                                "aten::_foreach"
+                            and (
+                                n.target._schema.name.startswith(
+                                    "aten::_foreach"
+                                )
+                                or n.target._schema.name.startswith(
+                                    "aten::_fused_adam"
+                                )
```
Comment (fegin) [inline torch/distributed/_spmd/distribute.py:397]: nit: startswith() accepts a tuple of string.

**Review by fegin — APPROVED** [review]
Nice!

File: torch/distributed/_spmd/distribute.py (hunk 2 above)
Comment (mrshenli) [reply torch/distributed/_spmd/distribute.py:397]: Thanks! Let me add this change on top.

**Comment by mrshenli** [comment]
Test failure is irrelevant

```
test_jit_fuser_te.py::TestTEFuserDynamic::test_dynamic_shapes Command took >20min, returning 124
FINISHED PRINTING LOG FILE of test_jit_fuser_te 1/3 (/var/lib/jenkins/workspace/test/test-reports/test_jit_fuser_te_e1f8106l.log)

test_jit_fuser_te 1/3 failed!
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/run_test.py", line 1398, in <module>
    main()
  File "/var/lib/jenkins/workspace/test/run_test.py", line 1356, in main
    raise RuntimeError(
RuntimeError: test_jit_fuser_te 1/3 failed!
```

**Pull request closed without merging**